src/components/tl/ucp/alltoall/alltoall_onesided.c

```diff
@@ -11,55 +11,306 @@
+    return;
 }
 
-void ucc_tl_ucp_alltoall_onesided_progress(ucc_coll_task_t *ctask)
+void ucc_tl_ucp_alltoall_onesided_put_progress(ucc_coll_task_t *ctask)
 {
-    ucc_tl_ucp_task_t *task  = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
-    ucc_tl_ucp_team_t *team  = TASK_TEAM(task);
-    ucc_rank_t         gsize = UCC_TL_TEAM_SIZE(team);
-    long *             pSync = TASK_ARGS(task).global_work_buffer;
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ptrdiff_t          src       = (ptrdiff_t)TASK_ARGS(task).src.info.buffer;
+    ptrdiff_t          dest      = (ptrdiff_t)TASK_ARGS(task).dst.info.buffer;
+    ucc_rank_t         grank     = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         gsize     = UCC_TL_TEAM_SIZE(team);
+    uint32_t           ntokens   = task->alltoall_onesided.tokens;
+    int64_t            npolls    = task->alltoall_onesided.npolls;
+    int64_t            cpolls    = 0;
+    ucc_mem_map_mem_h  src_memh  = TASK_ARGS(task).src_memh.local_memh;
+    ucc_mem_map_mem_h *dst_memh  = TASK_ARGS(task).dst_memh.global_memh;
+    uint32_t          *posted    = &task->onesided.put_posted;
+    uint32_t          *completed = &task->onesided.put_completed;
+    ucc_rank_t         peer      = (grank + *posted + 1) % gsize;
+    size_t             nelems;
 
-    if (ucc_tl_ucp_test_onesided(task, gsize) == UCC_INPROGRESS) {
-        return;
+    nelems = TASK_ARGS(task).src.info.count;
+    nelems = (nelems / gsize) * ucc_dt_size(TASK_ARGS(task).src.info.datatype);
+
+    for (; *posted < gsize; peer = (peer + 1) % gsize) {
+        UCPCHECK_GOTO(
+            ucc_tl_ucp_put_nb(PTR_OFFSET(src, peer * nelems),
+                              PTR_OFFSET(dest, grank * nelems), nelems,
+                              peer, src_memh, dst_memh, team, task),
+            task, out);
+
+        if (!alltoall_onesided_handle_completion(task, posted, completed,
+                                                 ntokens, npolls, peer,
+                                                 &cpolls)) {
+            return;
+        }
     }
 
-    pSync[0]           = 0;
-    task->super.status = UCC_OK;
+    alltoall_onesided_wait_completion(task, npolls);
+out:
+    return;
+}
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_get_start(ucc_coll_task_t *ctask)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team = TASK_TEAM(task);
+
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+}
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_put_start(ucc_coll_task_t *ctask)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team = TASK_TEAM(task);
+
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+}
```

Comment:
this is exactly the same as get_start, why not to keep just one function e.g. ucc_tl_ucp_alltoall_onesided_start?